### PR TITLE
PERF Removed classes instantiations and fetch to avoid multiple DB queries per invoice

### PR DESCRIPTION
**Pull request by Daviid-P**
# NEW Remove unnecessary queries
Removed classes instantiations and fetch to avoid multiple DB queries per invoice

### CODE REVIEW

**Comment by Daviid-P** [comment]
Created pull request as draft for #25615

File: htdocs/compta/prelevement/class/bonprelevement.class.php [hunk 1]
```diff
@@ -895,20 +895,17 @@ public function create($banque = 0, $agence = 0, $mode = 'real', $format = 'ALL'
 			$sql .= ", pfd.code_banque, pfd.code_guichet, pfd.number, pfd.cle_rib";
 			$sql .= ", pfd.amount";
 			$sql .= ", s.nom as name";
+			$sql .= ", f.ref, sr.bic, sr.iban_prefix, sr.frstrecur";
 			if ($type != 'bank-transfer') {
 				$sql .= " FROM ".MAIN_DB_PREFIX."facture as f";
+				$sql .= " LEFT JOIN " . MAIN_DB_PREFIX . "prelevement_facture_demande as pfd ON f.rowid = pfd.fk_facture";
```
Comment (eldy) [inline htdocs/compta/prelevement/class/bonprelevement.class.php:901]: Starting with v18, the table prelevement_facture_demande has been renamed into prelevement_demande

File: htdocs/compta/prelevement/class/bonprelevement.class.php [hunk 2]
```diff
@@ -895,20 +895,17 @@ public function create($banque = 0, $agence = 0, $mode = 'real', $format = 'ALL'
...
 			} else {
 				$sql .= " FROM ".MAIN_DB_PREFIX."facture_fourn as f";
+				$sql .= " LEFT JOIN " . MAIN_DB_PREFIX . "prelevement_facture_demande as pfd ON f.rowid = pfd.fk_facture_fourn";
 			}
-			$sql .= ", ".MAIN_DB_PREFIX."societe as s";
-			$sql .= ", ".MAIN_DB_PREFIX."prelevement_demande as pfd";
+			$sql .= " LEFT JOIN " . MAIN_DB_PREFIX . "societe as s ON s.rowid = f.fk_soc";
```
Comment (eldy) [inline htdocs/compta/prelevement/class/bonprelevement.class.php:906]: Why a left join on societe, link s.rowid = f.fk_soc is guaranted by foreign key.

File: htdocs/compta/prelevement/class/bonprelevement.class.php [hunk 3]
```diff
@@ -895,20 +895,17 @@ public function create($banque = 0, $agence = 0, $mode = 'real', $format = 'ALL'
...
+			$sql .= " LEFT JOIN " . MAIN_DB_PREFIX . "societe_rib as sr ON s.rowid = sr.fk_soc AND sr.default_rib = 1";
```
Comment (eldy) [inline htdocs/compta/prelevement/class/bonprelevement.class.php:907]: Note that 1 thirdparty can have 2 default value in this table, for example 1 for the 'ban' type and another one for the 'card' type.

Comment (Daviid-P) [reply htdocs/compta/prelevement/class/bonprelevement.class.php:907]: Fixed

File: htdocs/compta/prelevement/class/bonprelevement.class.php (hunk 1 above)
Comment (Daviid-P) [reply htdocs/compta/prelevement/class/bonprelevement.class.php:901]: Fixed

File: htdocs/compta/prelevement/class/bonprelevement.class.php (hunk 2 above)
Comment (Daviid-P) [reply htdocs/compta/prelevement/class/bonprelevement.class.php:906]: Whenever I do JOINs I always default to LEFT JOIN, I don't know why.

Should I change it to:
```php
$sql .= ", f.ref, sr.bic, sr.iban_prefix, sr.frstrecur";
if ($type != 'bank-transfer') {
	$sql .= " FROM ".MAIN_DB_PREFIX."facture as f";
	$sql .= " LEFT JOIN " . MAIN_DB_PREFIX . "prelevement_demande as pfd ON f.rowid = pfd.fk_facture";
} else {
	$sql .= " FROM ".MAIN_DB_PREFIX."facture_fourn as f";
	$sql .= " LEFT JOIN " . MAIN_DB_PREFIX . "prelevement_demande as pfd ON f.rowid = pfd.fk_facture_fourn";
}
$sql .= " JOIN " . MAIN_DB_PREFIX . "societe as s ON s.rowid = f.fk_soc";
$sql .= " LEFT JOIN " . MAIN_DB_PREFIX . "societe_rib as sr ON s.rowid = sr.fk_soc AND sr.default_rib = 1";
```

???

File: htdocs/compta/prelevement/class/bonprelevement.class.php (hunk 1 above)
Comment (Daviid-P) [reply htdocs/compta/prelevement/class/bonprelevement.class.php]: Should I change `pfd` to `pd` or `pld` ???

Comment (eldy) [reply htdocs/compta/prelevement/class/bonprelevement.class.php]: "pfd" into "pd" would be perfect but not so important

File: htdocs/compta/prelevement/class/bonprelevement.class.php (hunk 2 above)
Comment (eldy) [reply htdocs/compta/prelevement/class/bonprelevement.class.php:906]: when joinis garanted to match (no empty lines into the second table), we must use JOIN and NOT LEFT JOIN. Result would be the same but database can use perf optimization with JOIN not possible with LEFT JOIN.
So JOIN is better when we can.
Yes, suggested code seems good. Need to be tested however...

File: htdocs/compta/prelevement/class/bonprelevement.class.php (hunk 1 above)
Comment (Daviid-P) [reply htdocs/compta/prelevement/class/bonprelevement.class.php]: Done

**Pull request merged**